src/ucp/core/ucp_am.c

```diff
@@ -470,43 +479,86 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
+    ucp_am_hdr_t *hdr        = ucs_alloca(hdr_length);
 
-    ucp_am_fill_header(&reply_hdr.super, req);
-    reply_hdr.ep_ptr = ucp_request_get_dest_ep_ptr(req);
+    return ucp_am_zcopy_single_common(self, hdr, sizeof(*hdr), user_hdr_length,
+                                      UCP_AM_ID_SINGLE);
+}
 
-    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE_REPLY,
-                                  &reply_hdr, sizeof(reply_hdr),
-                                  ucp_proto_am_zcopy_req_complete);
+static ucs_status_t ucp_am_zcopy_single_reply(uct_pending_req_t *self)
+{
+    ucp_request_t *req            = ucs_container_of(self, ucp_request_t,
+                                                     send.uct);
+    unsigned user_hdr_length      = req->send.msg_proto.am.header_length;
+    size_t hdr_length             = sizeof(ucp_am_reply_hdr_t) +
+                                    user_hdr_length;
+    ucp_am_reply_hdr_t *reply_hdr = ucs_alloca(hdr_length);
+
+    reply_hdr->ep_ptr             = ucp_request_get_dest_ep_ptr(req);
+
+    return ucp_am_zcopy_single_common(self, (ucp_am_hdr_t*)reply_hdr,
+                                      sizeof(*reply_hdr), user_hdr_length,
+                                      UCP_AM_ID_SINGLE_REPLY);
 }
 
 static ucs_status_t ucp_am_zcopy_multi(uct_pending_req_t *self)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_am_first_hdr_t first_hdr;
+    ucp_am_first_hdr_t *first_hdr;
     ucp_am_mid_hdr_t mid_hdr;
+    size_t hdr_length;
+    ucs_status_t status;
+
+    if (req->send.state.dt.offset != 0) {
+        ucp_am_fill_middle_header(&mid_hdr, req);
+        return ucp_do_am_zcopy_multi(self, UCP_AM_ID_FIRST, UCP_AM_ID_MIDDLE,
+                                     NULL, 0ul, &mid_hdr, sizeof(mid_hdr),
+                                     ucp_proto_am_zcopy_req_complete, 1);
+    }
+
+    hdr_length = sizeof(*first_hdr) + req->send.msg_proto.am.header_length;
+    first_hdr  = ucs_alloca(hdr_length);
 
-    ucp_am_fill_first_header(&first_hdr, req);
-    ucp_am_fill_middle_header(&mid_hdr, req);
+    ucp_am_fill_first_header(first_hdr, req);
+    ucp_am_pack_user_header(first_hdr + 1, req);
 
-    return ucp_do_am_zcopy_multi(self, UCP_AM_ID_FIRST, UCP_AM_ID_MIDDLE,
-                                 &first_hdr, sizeof(first_hdr), &mid_hdr,
-                                 sizeof(mid_hdr), ucp_proto_am_zcopy_req_complete,
-                                 1);
+    status = ucp_do_am_zcopy_multi(self, UCP_AM_ID_FIRST, UCP_AM_ID_MIDDLE,
+                                   first_hdr, hdr_length, NULL, sizeof(mid_hdr),
+                                   ucp_proto_am_zcopy_req_complete, 1);
+    if (status != UCS_ERR_NO_RESOURCE) {
+        ucs_assert(UCS_STATUS_IS_ERR(status) ||
+                   (req->send.state.dt.offset != 0));
+        req->send.length -= req->send.msg_proto.am.header_length;
```

Comment:
why in zcopy single we update length before and rever if failed, and in zcopy multi update length after?

Reply:
will align with single zcopy